### DISCUSSION TITLE
Why would one install get "sqlalchemy.exc.InvalidRequestError: SQL expression, column, or mapped entity expected - got ''

### DISCUSSION THREAD

**Discussion by jeabraham** [sqlalchemy/sqlalchemy#10919] [Usage Questions]
I have a plugin in QGIS that uses SQL Alchemy. To avoid version problems, we're targeting the long term QGIS release which has bundled SQL Alchemy 1.3.24.

I have problem where one computer complains

sqlalchemy.exc.InvalidRequestError: SQL expression, column, or mapped entity expected - got ''

when I create a query on a Reflected 

q = session.query(ProjectCF)

Where ProjectCF is a DeferredReflection class.  

When I reflect and prepare, the table names are there.  

Reflected.prepare(self._engine)
insp = inspect(self._engine)
print(insp.get_table_names(schema="conformity_forecast"))

I've deployed the exact same code on two other computers and it's been working.  I cannot for the life of me figure out why one of three computers wouldn't work.  

Also, SQL Alchemy queries worked on this computer from this plugin before.  It seems something broke. 

What would be the next debugging step?  Is there something I can log in .prepare(engine) ?

Much thanks,

**Comment by CaselIT** [comment]
Hi,

The provided information is not much. Ensure that the correct sqlalchemy version is used in all systems

> What would be the next debugging step?

I would suggest reinstalling the env where you have an issue to rule out bad packages for some reason.

Also the complete stack trace would help provide more information

**Comment by jeabraham** [comment]
I made some progress.  

I had a class "Reflected" which was a subclass of DeferredReflection, and all my classes were subclasses of both Reflected and Base

class Reflected(DeferredReflection):
    __abstract__ = True

class MyActualClass(Reflected, Base):
  __tablename__ = "my_table"
...

In the prepare method in sqlalchemy.ext.declarative.api.py there's a loop over classes_for_base

        to_map = _DeferredMapperConfig.classes_for_base(cls)

and this wasn't looping over my sub-sub classes in the one installation. Perhaps a bug in classes_for_base(cls) ? But all the installations are using the same SQL Alchemy versions 1.3.24

I don't recall the reason for setting up the class Reflected.  I tried removing it, and now at least it's trying to initialize all the classes.  Now fails because some relationships are being defined before both dependent subclasses are defined.

**Comment by jeabraham** [comment]
After further investigation we discovered that the version that worked wasn't actually on 1.3.24.  It was on 2.0.23.  It seems there were some bugs in 1.3.24 for the reflection and relationships that don't exist anymore in 2.0.23.

It's too bad QGIS is bundled with such an old version of SQL Alchemy.  There's often problems installing 2.0.23 inside of QGIS, but it seems those problems are small in comparison to this functionality bug in 1.3.24. 

Thanks for listening,
